Issues/variants support through url

mohitnatoo opened the pull request:
Changes for the fix of Issue #18818 .
- This is continuation from what has been left [here](https://github.com/rails/rails/pull/19245).
- The branch has been rebased with master as requested [here](https://github.com/rails/rails/pull/19245#issuecomment-92059195) and few modifications to the already existing code have been done accordingly.

r? @kaspth

vipulnsward commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+*   Allow variants to be set through the URL
+
+    request.variant can be set from url. For example, accessing
+    http://example.com/posts.html+partial set `request.variant = [:partial]`
```
`sets`

vipulnsward commented on actionpack/CHANGELOG.md:
```diff
@@ -373,6 +381,10 @@
+
+*   Drop request class from RouteSet constructor.
+
+    If you would like to use a custom request class, please subclass and implemet
```
remove this.

vipulnsward commented on actionpack/lib/action_dispatch/http/mime_negotiation.rb:
```diff
@@ -156,7 +160,7 @@ def negotiate_mime(order)
       def valid_accept_header
         (xhr? && (accept.present? || content_mime_type)) ||
-          (accept.present? && accept !~ BROWSER_LIKE_ACCEPTS)
+            (accept.present? && accept !~ BROWSER_LIKE_ACCEPTS)
```
ditto, no change here.

vipulnsward commented on actionpack/lib/action_dispatch/http/mime_negotiation.rb:
```diff
@@ -95,7 +95,11 @@ def variant=(variant)
       def variant
-        @variant ||= ActiveSupport::ArrayInquirer.new
+        @variant ||= if parameters.key? :variant
+                       send(:variant=, parameters[:variant].try(:to_sym))
```
`public_send`

kaspth commented on actionpack/lib/action_dispatch/http/mime_negotiation.rb:
```diff
@@ -95,7 +95,11 @@ def variant=(variant)
       def variant
-        @variant ||= ActiveSupport::ArrayInquirer.new
+        @variant ||= if parameters.key? :variant
+                       public_send(:variant=, parameters[:variant].try(:to_sym))
```
Can't we use `self.variant =` here? We also try to avoid `try` internally in the framework, let's remove it by checking for a value in `parameters[:variant]` instead of there being a key. Like this:

```ruby
def variant
  @variant ||=
    if parameters[:variant]
      self.variant = parameters[:variant].to_sym
    else
      ActiveSupport::ArrayInquirer.new
    end
end

dhh commented:
Eager to see this be part of 5.0. Thanks for working on it!

mohitnatoo commented:
@kaspth made the review changes.

kaspth commented:
@mohitnatoo thanks. Remember to add yourself to the commit message in a format that Rails contributors understands: http://contributors.rubyonrails.org/faq. Otherwise you won't get credit for your contribution! :)

About the test failures. In the majority of the Action Pack failures, it seems you need to add more spaces to `Controller#Action` headers so they match their new positions. These tests also seem to break: https://github.com/rails/rails/blob/a00c36feea6c0271b5ad48a949ef294514fdef52/actionpack/test/journey/routes_test.rb#L26-L43.

kaspth commented:
The Railties failures are also because of `Controller#Action` not having the appropriate amount of spaces before them.

prathamesh-sonpatki commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+*   Allow variants to be set through the URL
+
```
Add `.` at the end of statement.

prathamesh-sonpatki commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+*   Allow variants to be set through the URL
+
+    request.variant can be set from url. For example, accessing
+    http://example.com/posts.html+partial sets `request.variant = [:partial]`
```
Change it to `request.variant`, url -> URL

mohitnatoo committed:
Allow variants to be set through the URL

`:variant` is tied to `:format` like (.:format(+:variant)),
and constraints of :format is set to `[^\+]+` rather than `.+`.
`request.variant` is set to `:variant` parameter default.
For example, accessing http://example.com/posts.html+partial set
`request.variant = [:partial]` automatically if route is
`posts(.:format(+:variant))`.

- Commit for fix #18818 which is referenced from #19245.

- Fix for #18818 which is by taking #19245 as base

- fixing changelog.md rebase conflicts

- added change to changelog.md

- made review changes

- check value of variant before symbolizing.

- corrected spaces in test cases assertions in inspector_test

- corrected spaces in test cases assertions in rake_test.rb

- corrected test case to have format(+:variant)

- corrected test case to have format(+:variant)

-updating changelog.md

- fixed failing test cases due to spacing.

- fixed failing test cases due to spacing.

- Modified normalize_path in mapper.rb to have only :format instead of adding variant as well.

- Reverting previous commit of change made in mapper.rb

- updated test expectation after rebasing branch with master

- fixing failing railties tests.

- fixing failing railties tests.

vipulnsward commented on actionpack/CHANGELOG.md:
```diff
@@ -10,6 +10,14 @@
+    http://example.com/posts.html+partial sets `request.variant = [:partial]`
+    automatically. `:variant` is tied to `:format` like `(.:format(+:variant))`.
+
+    *Hong ChulJu*, *Mohit Natoo*
```
New entries goes to the top of the file.

kaspth commented on railties/test/application/rake_test.rb:
```diff
@@ -129,8 +129,8 @@ def test_rake_routes_calls_the_route_inspector
 
-      output = Dir.chdir(app_path){ `bin/rails routes` }
-      assert_equal "Prefix Verb URI Pattern     Controller#Action\n  cart GET  /cart(.:format) cart#show\n", output
+      output = Dir.chdir(app_path){ `bin/rake routes` }
```
Should be `bin/rails routes`, like it was.

kaspth commented:
Railties is failing because of the updates we made to `bin/rails routes` elsewhere, so there are more tests to add spaces in.

Action Pack seems to be failing in the router tests, which likely don't expect the new default of `(+variant)`. Can you check those out too?

dhh commented:
Bump.

kaspth commented:
Punted to 5.1 because of internal discussions where we concluded we need more time to evaluate changing the meaning of `+` in generated URLs.

cc @prathamesh-sonpatki @pixeltrix

mikegee commented:
Has there been additional discussion on this?
